include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -3,150 +3,237 @@
 
 #include <irr/builtin/material_compiler/glsl/common.glsl>
 
-void instr_eval_execute(in instr_t instr, in vec3 L)
+void instr_eval_execute(in instr_t instr, in MC_precomputed_t precomp, inout irr_glsl_LightSample s, inout irr_glsl_AnisotropicMicrofacetCache _microfacet, in bool skip)
 {
-	uint op = instr_getOpcode(instr);
-
-	//speculative execution
-	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
-	params_t params = instr_getParameters(instr, bsdf_data);
-	float bxdf_eval_scalar_part;
-	uint ndf = instr_getNDF(instr);
-	float a = params_getAlpha(params);
-	float a2 = a*a;
-#ifndef ALL_ISOTROPIC_BXDFS
-	float ay = params_getAlphaV(params);
-	float ay2 = ay*ay;
-#endif
+	const uint op = instr_getOpcode(instr);
+	const bool is_bxdf = op_isBXDF(op);
+	const bool is_bsdf = !op_isBRDF(op); //note it actually tells if op is BSDF or BUMPMAP or SET_GEOM_NORMAL (divergence reasons)
+	const float cosFactor = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
+	const bool positiveCosFactor = cosFactor > FLT_MIN;
+	const bool is_bxdf_or_combiner = op_isBXDForCoatOrBlend(op);
 
-	float cosFactor = op_isBSDF(op) ? abs(currBSDFParams.isotropic.NdotL):max(currBSDFParams.isotropic.NdotL,0.0);
+	uvec3 regs = instr_decodeRegisters(instr);
+	mat2x3 ior;
+	mat2x3 ior2;
+	params_t params;
+	irr_glsl_AnisotropicMicrofacetCache microfacet;
+	bsdf_data_t bsdf_data;
 
-	if (cosFactor>FLT_MIN && op_hasSpecular(op))
+	const bool run = !skip;
+
+	if (run && is_bxdf_or_combiner)
```

Comment:
I'd fold `positiveCosFactor` into `run` and make an `II` for `is_combiner` (so COAT and BLEND fetch the params, ior and bsdf data)

Reply:
just be careful because `positiveCosFactor` is now folded into `run` and unless you're able to tell if your COAT coats a DIFFTRANS or BLEND with at least one BSDF as a child, you need to fetch the data

Reply:
still a TODO

Reply:
move intitialization of `is_valid` from here
https://github.com/Crisspl/IrrlichtBAW/blob/mitsuba/include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl#L13

to here (and initialize to true) https://github.com/Crisspl/IrrlichtBAW/blob/mitsuba/include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl#L42

only use if you have `OP_DIELECTRIC` defined

(Current usage confuses me and probably the compiler as well)

Reply:
> I'd fold `positiveCosFactor` into `run` and make an `II` for `is_combiner` (so COAT and BLEND fetch the params, ior and bsdf data)

this is only relevant is COAT or BLEND are considered BRDFs by you (`is_bsdf` used for clamping `NdotV` and `NdotL` could be false in these cases)

Reply:
also on line https://github.com/Crisspl/IrrlichtBAW/blob/mitsuba/include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl#L62

its safe to use the `wo_clamps` variant